[String] Last-minute ABI adjustment: 4-bit discriminator

milseman opened the pull request:
Switch ABI to only use 4 discriminator bits.
    
In anticipation of potential future HW features, e.g. armv8.5 memory
tagging, only use the high 4 bytes as discriminator bits in
_BridgeObject rather than the top 8 bits. Utilize two perf flags to
cover this instead. This requires shifting around a fair amount of
internal complexity.

This includes a sizable refactoring of 32-bit support, making this kind of adjustment possible.

milseman committed:
[String] Refactor 32-bit StringObject.

Remove Discriminator, Flags, etc., abstractions from
StringObject. These cause code divergence between 32-bit and 64-bit
ABI, complicate ABI changes, and otherwise contribute to bloat.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please benchmark

milseman commented:
@swift-ci please test

lorentey reviewed (approved):
Nice simplifications!

milseman commented:
Trying an asserts build locally to repro. Seems to just be a few `_internalInvariant` checks that need updating.

milseman commented:
@swift-ci please test

milseman committed:
[String] Switch ABI to only use 4 discriminator bits.

In anticipation of potential future HW features, e.g. armv8.5 memory
tagging, only use the high 4 bytes as discriminator bits in
_BridgeObject rather than the top 8 bits. Utilize two perf flags to
cover this instead. This requires shifting around a fair amount of
internal complexity.

milseman commented:
@lorentey can you take a look? In addition to the refactoring, this has ABI changes to only use 4-bit discriminators.

Catfish-Man commented on stdlib/public/core/SmallString.swift:
```diff
@@ -105,13 +105,9 @@ extension _SmallString {
-        self._storage.0,
-        self._storage.1 & _StringObject.Nibbles.largeAddressMask)
-    }
+    return (self._storage.0, self._storage.1 & 0x00FF_FFFF_FFFF_FFFF)
```
Losing the nice explanatory name on the mask here makes me a bit sad

milseman replied:
Right, but that name was misleading and conflating. This is a mask only for the small-string contents, which no longer aligns with addresses.

I thought about adding a static property for this, which would be inlinable and in our ABI, and only used in this one place. I'm not sure how to weigh the "at least it has an name" with "one level of cognitive indirection" if there's only one use site.

stephentyrone replied:
A local name could help readability here. `let smallStringCodeUnitsMask = 0x00FF_FFFF_FFFF_FFFF` or whatever. An actual static property is probably overkill.

Catfish-Man commented on stdlib/public/core/StringStorage.swift:
```diff
@@ -507,7 +510,7 @@ extension _StringStorage {
   private func _postRRCAdjust(newCount: Int, newIsASCII: Bool) {
     let countAndFlags = CountAndFlags(
-      count: newCount, isASCII: newIsASCII)
+      mortalCount: newCount, isASCII: newIsASCII)
```
"mortalCount" is the most metal variable name I've seen in a while

milseman replied:
🤘

Catfish-Man commented on stdlib/public/core/StringBridge.swift:
```diff
@@ -281,9 +281,13 @@ extension String {
       }
     }
     if _guts._object.isImmortal {
+      // TODO: We'd rather emit a valid ObjC object statically than create a
```
Do we have a bug tracking this? I should either CC on it or steal it if so

milseman replied:
If we do, then I assigned it to you ;-)

stephentyrone commented on stdlib/public/core/SmallString.swift:
```diff
@@ -108,27 +99,21 @@ extension _SmallString {
-        self._storage.0,
-        self._storage.1 & _StringObject.Nibbles.largeAddressMask)
-    }
+    return (self._storage.0, self._storage.1 & 0x00FF_FFFF_FFFF_FFFF)
```
Is this mask correct, or should it be `0x0FFF_` ...? Is there a reason not to use the named mask any longer here?

milseman replied:
Small string's contents have now diverged from the addresses. Small string uses the second-highest nibble for the count, while large strings consider that to be part of the address (or whatever future HW or OSes want).

stephentyrone commented on stdlib/public/core/SmallString.swift:
```diff
@@ -108,27 +99,21 @@ extension _SmallString {
-  @inlinable
   internal func computeIsASCII() -> Bool {
-    // TODO(String micro-performance): Evaluate other expressions, e.g. | first
     let asciiMask: UInt64 = 0x8080_8080_8080_8080
```
I would prefer to see this named constant used uniformly everywhere, but I'm OK with keeping it out of this PR.

milseman replied:
Right, if it's in more than one place I agree. Otherwise, not sure about the tradeoff of baking it into our ABI and having an extra level of indirection, unless there is a lot of semantic value in the name.

milseman replied:
Err, never mind on the ABI part, as this isn't inlined.

stephentyrone replied:
Definitely don't think we need it to be anywhere in the ABI =)

Catfish-Man commented on stdlib/public/core/StringBridge.swift:
```diff
@@ -281,9 +281,13 @@ extension String {
         immortal: _guts._object.fastUTF8.baseAddress!,
-        countAndFlags: _guts._object._countAndFlags)
+        countAndFlags: _StringObject.CountAndFlags(
+          sharedCount: _guts.count, isASCII: gutsCountAndFlags.isASCII))
```
what's the reasoning for making a new CountAndFlags here rather than using gutsCountAndFlags? Just Shared Strings needing some different flags?

milseman replied:
Shared strings do not have tail-allocated contents, which is one of the new perf flags, so we need to strip it. I very slightly prefer using initializers than mutation, because it makes coding style more uniform for things like `_invariantCheck()` not being called on temporarily-invalid states.

milseman commented:
https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx-perf/736/

jrose-apple commented on stdlib/public/core/SmallString.swift:
```diff
@@ -231,14 +216,28 @@ extension _SmallString {
+  internal init(leading: UInt64, trailing: UInt64, count: Int) {
+    _internalInvariant(count <= _SmallString.capacity)
+
+    let isASCII = (leading | trailing) & 0x8080_8080_8080_8080 == 0
```
Ooh, nice. (I know it's not new; I just hadn't seen it before.)

jrose-apple replied:
Actually, given how fast this is, I wonder if the dedicated `isASCII` bit is even needed, or if it's worth saving for something.

milseman replied:
It's not that fast; we have to mask off the discriminator too. While ARM can encode the constant as an immediate, but it's still more code. Specifically for small strings, ASCII-ness is by far the most common and important characteristic to know.

The other contender for this bit would be remembering both isASCII and doesn't have CR+LF, i.e. single-scalar-grapheme. But, I think it's better to check that on Character access.

```swift
  public func _forCodeGen_1() { if isASCII { foo() } else { bar() } }
  public func _forCodeGen_2() { if computeIsASCII() { foo() } else { bar() } }
```

```
_$ss12_SmallStringV13_forCodeGen_1yyF:
  tbnz       x1, 0x3e, loc_15c474

  b          _$ss12_SmallStringV3baryyFTf4d_n

loc_15c474:
  b          _$ss12_SmallStringV3fooyyFTf4d_n



_$ss12_SmallStringV13_forCodeGen_2yyF:
  orr        x8, xzr, #0x8080808080808080
  movk       x8, #0x80, lsl #48
  and        x8, x1, x8
  and        x9, x0, #0x8080808080808080
  orr        x8, x8, x9
  cbz        x8, loc_15c494

  b          _$ss12_SmallStringV3baryyFTf4d_n
loc_15c494:
  b          _$ss12_SmallStringV3fooyyFTf4d_n
```

Considering that this check should probably be at the top of every String API inlined into user code, it's probably worth it.

jrose-apple replied:
In theory that could be

```
mov x8, x0, lsl #8
orr x8, x1, x8
and x8, x8, #0x8080808080808080
cbz x8, loc
```

but I take your point.

jrose-apple commented on stdlib/public/core/StringObject.swift:
```diff
@@ -64,17 +60,22 @@ internal struct _StringObject {
   as above, except bit 7 is omitted from storage -- it is left free, to supply
   extra inhabitants in the StringObject structure. The missing bit can be
   recovered by looking at `_variant.isImmortal`.
```
This part of the comment is no longer up to date, since there are not 7 bits.

jrose-apple commented on stdlib/public/core/StringObject.swift:
```diff
@@ -820,102 +605,166 @@ extension _StringObject {
+ NOTE: isNativelyStored is *specifically* allocated to b61 to align with the
+ bit-position of isSmall, which allows us to check for native storage without
+ an extra branch checking small-ness. Otherwise, these bits are invalid to read
+ from small strings.
```
I don't quite understand this bit. Aren't they in different words?

jrose-apple replied:
Gotcha! Clever.

jrose-apple commented on stdlib/public/core/StringObject.swift:
```diff
@@ -820,102 +605,166 @@ extension _StringObject {
+   - `largeAddressBits` holds an instance of `_StringStorage`.
+ isTailAllocated: start of the code units is at the stored address + `nativeBias`
+   - `isNativelyStored` always implies `isTailAllocated`, but not vice versa
+      (e.g. literals)
```
Do we not have a canonical form for "I point directly to unbiased string data" yet?

milseman replied:
We don't have anything that directly (unbiased) points to contents.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please benchmark

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please benchmark

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please benchmark

milseman commented:
So yeah, need to update LLDB

milseman commented:
Please test with following pull request:
https://github.com/apple/swift-lldb/pull/1168

@swift-ci please test
